apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1.IntegrationTests/Emulator/metadata_server.py

```diff
@@ -0,0 +1,366 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import OrderedDict
+from flask import Flask, request
+from werkzeug.wrappers import Request
+import common, errno, getopt, json, logging, os, socket, sys, time
+
+numeric_project_id = 12345;
+numeric_project_id_token = u'<NUMERIC_PROJECT_ID>';
+project_id = u'fake-project';
+project_id_token = u'<PROJECT_ID>';
+
+app = Flask(__name__)
+
+def excludeKeyInRecursiveGet(key):
+  return key == u'token'
+
+def formatResponse(data, status_code, content_type):
+  return data, status_code, {common.content_type: content_type}
+
+def formatOrderedDictContents(data):
+  def getKeyDescription(key):
+    if not excludeKeyInRecursiveGet(key):
+      if isinstance(data[key], (IndexedList, OrderedDict)):
+        return key + u'/\n'
+
+    return key + u'\n'
+
+  return ''.join(getKeyDescription(key) for key in data)
+
+class HeaderCheckMiddleware(object):
+  def __init__(self, app):
+     self.app = app
+
+  def __call__(self, environ, start_response):
+    req = Request(environ, shallow=True)
+    if (req.headers.get(common.metadata_flavor) != common.metadata_flavor_google):
+      start_response('403 Forbidden',
+                     [(common.content_type, common.content_type_html),
+                      (common.metadata_flavor, common.metadata_flavor_google),
+                      (common.server, common.server_value)])
+      return ['Missing Metadata-Flavor:Google header.']
+    return self.app(environ, start_response)
+
+# Collection whose content is accessed via an index value in the relative URL.
+class IndexedList:
+  def __init__(self, items):
+    self._items = items
+
+  def __iter__(self):
+    return self._items.__iter__()
+
+  def __len__(self):
+    return self._items.__len__()
+
+  def __delitem__(self, key):
+    self._items.__delitem__(key)
+
+  def __getitem__(self, key):
+    return self._items.__getitem__(key)
+  
+  def __setitem__(self, key, value):
+    self._items.__setitem__(key, value)
+
+  def formatContents(self):
+    return ''.join(unicode(i) + u'/\n' for i in range(len(self._items)))
+
+class MetadataJSONEncoder(json.JSONEncoder):
+  def ensureOrderedDict(self, o):
```

Comment:
What types does this assume? Just OrderedDicts that need to be fixed? If you get an IndexedList, do you need to convert it or is it ok? (I think it should be ok since they have a stable order?)